### PR TITLE
Add doco for Router

### PR DESCRIPTION
**Pull request by SzymonPobiega**

### CODE REVIEW

**Comment by SzymonPobiega** [comment]
@ramonsmits @dvdstelt can you guys take a look at this? This is just for start. It does not cover the route table very well.

**Comment by SzymonPobiega** [comment]
@Particular/docs-maintainers can you review this?

File: nservicebus/router/index.md [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+title: Router
+summary: How to connect parts of the system that use different transports 
+component: Router
+related:
+- samples/azure/azure-service-bus-msmq-bridge
+- samples/msmq/sql-bridge
+reviewed: 2018-05-04
+---
+
+`NServiceBus.Router` is a universal component that connects parts of an NServiceBus-based solution that otherwise cannot talk to each other (e.g. because they use different transport or transport settings or because of physical barriers).
```
Comment (dvdstelt) [inline nservicebus/router/index.md]: > that otherwise cannot talk to each other (e.g. because they use different transport or transport settings

I think it should be

> that otherwise could not talk to each other (e.g. because they use different transports or transport settings

File: nservicebus/router/index.md [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+Regular endpoints connect to the router using *connectors* that allow them to configure the routing
+
+snippet: connector
+
+The snippet above tells the endpoint that a designated router listens on queue `MyRouter` and that messages of type `MyMessage` should be sent to the endpoint `Receiver` via the router. It also tell the subscription infrastructure that the event `MyEvent` is published by the endpoint `Publisher` that is hosted behind the router.
+
+
+## Router configuration
+
+The following snippet shows a simple MSMQ-to-RabbitMQ router configuration
```
Comment (dvdstelt) [inline nservicebus/router/index.md]: Maybe explain here how the router is hosted? It's unclear to a first-time reader if this is hosted on its own or inside an endpoint.

File: Snippets/Router/Router_2/API.cs [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+
+    public void OtherAPIs()
+    {
+        var routerConfig = new RouterConfiguration("MyRouter");
+
+        #region recoverability
+
+        routerConfig.CircuitBreakerThreshold = 20;
+        routerConfig.DelayedRetries = 10;
+        routerConfig.ImmediateRetries = 10;
```
Comment (dvdstelt) [inline Snippets/Router/Router_2/API.cs]: Nothing to do with the doco, but would it not be nicer if you could configure it exactly like [delayed retries](https://docs.particular.net/nservicebus/recoverability/configure-delayed-retries#configuring-delayed-retries)?

File: nservicebus/router/index.md [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+snippet: recoverability
+
+In addition to immediate and delayed retries, the router has built-in outage detection through a *circuit breaker*. After a number of consecutive failures the circuit breaker is triggered which causes the interface to enter the *throttled mode*. In this mode the interface processes a single message at a time and pauses after each processing attempt. The interface goes back to the normal mode after the first successful processing attempt. When in the throttled mode the router does not increment the delayed retries counter to prevent messages being sent to the poison message queue due to infrastructure outages.
+
+## Topologies
+
+A router consists of multiple [NServiceBus.Raw](/nservicebus/rawmessaging/) endpoints, called *interfaces* and a *routing protocol* that controls how messages should be forwarded between them.
```
Comment (dvdstelt) [inline nservicebus/router/index.md]: Why was this added? It's unclear how the reader benefits from this?

File: nservicebus/router/index.md [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
+
+### Bridge
+
+![Bridge](bridge.svg)
+
+The arrows show the path of messages sent from `Endpoint A` to `Endpoint C` and from `Endpoint D` to `Endpoint B`. The messages cross the router in the opposite direction. Each message is initially sent to the router queue (based on the routing configuration of the connector) and then forwarded to the destination queue. There is one additional *hop* compared to a direct communication of endpoints. The following snippet configures the built-in *static routing protocol* to forward messages between the router's interfaces.
```
Comment (dvdstelt) [inline nservicebus/router/index.md]: > The messages cross the router in the opposite direction.

I don't get this?

Comment (dvdstelt) [reply nservicebus/router/index.md]: > (based on the routing configuration of the connector)

I would probably remove this, doesn't add anything I think?

File: nservicebus/router/index.md [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+
+### Backplane
+
+![Backplane](backplane.svg)
+
+Two or more routers can be connected together to form a _backplane_ topology. This setup usually makes most sens for the geo-distributed systems. The following snippet configures the router hosted in the Europen part of the globally distributed system to route messages coming from outsize via the Azure Storage Queues interface directly to the local endpoints and to route messages sent by local endpoints to either East or West United States through *designated gateway* routers.
+
+NOTE: The *designated gateway* concept is not related to the NServiceBus.Gateway package. When the *designated gateway* is specified in the route, the message is forawrded to it instead of the actual destination.
+
+snippet: three-way-router
```
Comment (dvdstelt) [inline nservicebus/router/index.md]: This is the exact name snippet name as the one above? Is that correct?

File: nservicebus/router/index.md [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+
+NOTE: All three interfaces use the same transport type (SQL Server Transport) but may use different settings e.g. a different database instances. This way each part of the system (Sales, Shipping and Billing) can be autonomous and own its database server yet they still can exchange messages in the same way as if they were connected to a single shared instance.
+
+
+### Backplane
+
+![Backplane](backplane.svg)
+
+Two or more routers can be connected together to form a _backplane_ topology. This setup usually makes most sens for the geo-distributed systems. The following snippet configures the router hosted in the Europen part of the globally distributed system to route messages coming from outsize via the Azure Storage Queues interface directly to the local endpoints and to route messages sent by local endpoints to either East or West United States through *designated gateway* routers.
```
Comment (dvdstelt) [inline nservicebus/router/index.md]: > makes most sens

sense (with an additional 'e')

Comment (dvdstelt) [inline nservicebus/router/index.md]: > coming from outsize

outside

File: nservicebus/router/index.md [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+
+NOTE: The *designated gateway* concept is not related to the NServiceBus.Gateway package. When the *designated gateway* is specified in the route, the message is forawrded to it instead of the actual destination.
```
Comment (dvdstelt) [inline nservicebus/router/index.md]: > forawrded 

forwarded

File: nservicebus/router/index.md (hunk 6 above)
Comment (SzymonPobiega) [reply nservicebus/router/index.md]: Nope. A copy-paste error 👍

**Comment by SzymonPobiega** [comment]
Thanks a lot @dvdstelt ❤️ . I've fixed the issues.

**Comment by ramonsmits** [comment]
@SzymonPobiega 

Summary of my comments/questions as mentioned during our call:

- I assumed atleast-once delivery but you stated that transactional-exact-once is used when transports support it
- are multiple routes support to have a broadcast
- maybe add a section about message format that is doesn't support translating JSON to/from XML
- are subscriptions supported on transports that only support native pubsub
- could this be a replacement for the ASB multi-instance or how is this different?
- I associate back planes with broadcasts as used with SignalR but is seems it works different so maybe it is possible to clarify to avoid confusion.

**Comment by SzymonPobiega** [comment]
@dvdstelt can you approve?

**Comment by SzymonPobiega** [comment]
@ramonsmits I'll address you feedback in the next version and merge this as-is now. Thanks ❤️
